### PR TITLE
neteq preemptive expand improvements

### PR DESCRIPTION
**Pull request by adam-kiss-sg**
* fixed best_normalized_correlation index handling and position return
* slightly optimized best_normalized_correlation to better fit into f32
* use dynamic correlation threshold for preemptive expand
* allow for larger max_add value in preemptive expand to support deep male voices

## What is this PR about?

This PR fixes **critical bugs** and improves the **preemptive expand** operation, which stretches audio when the buffer is running low (but not yet empty). Think of it as the opposite of accelerate.

## Part 1: Fixed a Serious Math Bug in `best_normalized_correlation`

**The Bug:**
The correlation calculation was fundamentally broken:

```rust
// BEFORE - Wrong normalization!
corr2 = sum_ab^2 / (sum_a2 * sum_b2)  
// This squared the correlation, losing sign information
// Also used wrong energy formula
```

```rust
// AFTER - Proper normalization
corr = sum_ab / energy
// Where energy = max(x*x, y*y) for each sample pair
// Returns correlation in range [-1.0, 1.0] instead of [0, 1]
```

**Why this matters:** The old version couldn't detect **anti-correlation** (signals that are opposite). This caused it to treat perfectly mismatched audio (which should sound terrible when overlapped) the same as uncorrelated audio.

### Fixed Index Handling Bug
The original code had off-by-one errors in positioning the repeated audio segment. The new version correctly:
- Finds where to duplicate audio based on correlation
- Positions the crossfade at the right spot
- Copies regions in the correct order

### Dynamic Correlation Threshold (Similar to PR 475!)
Instead of accepting any correlation value, it now:

```rust
if best_corr < self.corr_threshold {
    self.corr_threshold *= 0.98  // Lower threshold, be less picky
    return NoStretch  // Don't stretch if quality would be bad
} else {
    self.corr_threshold = max(best_corr * 0.95)  // Raise threshold
}
```

This prevents stretching when there's no good match point, which would cause audible artifacts.

### Larger Expansion Range for Deep Voices
**Before:** Max expansion was limited to ~25% of frame  
**After:** Searches up to `(output.len() - overlap) / 2`

**Why:** Male voices can have pitch periods of ~12ms (85 Hz). The old code couldn't find the repetition period for deep voices, making preemptive expand fail for bass speakers!

## Part 3: Comprehensive Testing

Added extensive test cases for the correlation function:
- Perfect alignment (corr = 1.0)
- Anti-alignment (corr = -1.0)  
- Out-of-phase signals (corr ≈ 0)
- Zero signals (edge case)
- Different positions

## Part 4: Cleanup

Deleted the old `worker-example` directory (1000+ lines of outdated example code).

## Bottom Line

This PR transforms preemptive expand from **buggy and unreliable** to **mathematically correct and adaptive**. The correlation bug especially was causing it to make poor decisions about where to duplicate audio. Now it:
- Actually finds good match points
- Adapts to signal characteristics
- Works for all voice ranges (including deep voices)
- Refuses to stretch when it would sound bad

### CODE REVIEW

File: neteq/src/signal.rs [hunk 1]
```diff
@@ -42,35 +42,37 @@ pub fn best_normalized_correlation(a: &[f32], b: &[f32], len: usize) -> (usize,
...
-                sum_ab * sum_ab / (sum_a2 * sum_b2)
-            };
-
-            if corr2 > best_corr2 {
-                best_corr2 = corr2;
-                best_pos = i;
+        let x = a[i];
+        let y = b[i];
+        corr += x * y;
+        energy += (x * x).max(y * y);
```
[the remark this answers is not in the document]
Comment (adam-kiss-sg) [reply neteq/src/signal.rs:53]: It doesn't matter much in practice. Using max here will make perfect alignment result in a correlation of 1.0.
Alternative could be using avg: (x  * x + y * y)/2. 

Using max will make the alignment between a silent and loud sample less, which is preferable. And for samples that are well aligned and have similar energy the result will be similar with max or avg.

File: neteq/src/signal.rs [hunk 2]
```diff
@@ -42,35 +42,37 @@ pub fn best_normalized_correlation(a: &[f32], b: &[f32], len: usize) -> (usize,
...
+
+            if normalized_corr > best_corr {
+                best_corr = normalized_corr;
+                best_pos = i_start;
             }
+
+            let old_x = a[i_start];
+            let old_y = b[i_start];
+            corr -= old_x * old_y;
+            energy -= (old_x * old_x).max(old_y * old_y);
```
[the remark this answers is not in the document]
Comment (adam-kiss-sg) [reply neteq/src/signal.rs:71]: Not true.

**Comment by adam-kiss-sg** [comment]
Hi @darioalessandro !

I added some tests, and merged some others as test cases to a single table driven test.

normalized_correlation and best_normalized_correlation now work a bit differently:
The main difference is that normalized_correlation uses multiplication and square root, while best_normalized_correlation uses just the sum of max energies to to normalize. 

Reason for that change:
- wanted to avoid using sqrt too much because it's an expensive operation
- calculating with squares didn't fit properly into f32
- i did a lot of experiments with different calculations and normalizations (also tried squares with f64), and it didn't make much of a difference in practice, so I just wanted to stick with something that's computationally cheap

At this point, normalized_correlation is not used.

So, what do you think about the algorithm, and about the normalized_correlation function? 
Are you ok with keeping best_normalized_correlation as it is in this pr? 
Also, should I remove normalized_correlation, or keep it as is? (Or update it to also use max)?
And if I were to keep it, are you ok with me removing the test_best_correlation_consistency test? (I commented it out, since the two functions work differently now).

**Comment by darioalessandro** [comment]
Follow up, remove normalized_correlation

File: neteq/src/signal.rs [hunk 3]
```diff
@@ -124,27 +126,137 @@ mod tests {
         );
     }
 
+    // /// Test 2: Consistency Check
+    // /// The sliding window version should match manual calculation
+    // #[test]
+    // fn test_best_correlation_consistency() {
```
Comment (darioalessandro) [inline neteq/src/signal.rs:132]: follow up: burn down commented code

File: neteq/src/signal.rs [hunk 4]
```diff
@@ -124,27 +126,137 @@ mod tests {
...
+            want_best_pos: Option<usize>,
+            want_best_corr_min: f32,
+            want_best_corr_max: f32,
+        }
 
-        assert!(
-            (best_corr2 - manual_corr2).abs() < 0.001,
-            "Sliding window correlation should match manual calculation: {best_corr2} vs {manual_corr2}"
-        );
+        for test_case in [
```
Comment (darioalessandro) [inline neteq/src/signal.rs:165]: @adam-kiss-sg this is neat to cover all important scenarios!!
